rtt_ros2_params/src/rtt_ros2_params.cpp

```diff
@@ -0,0 +1,296 @@
+}
+
+bool Params::setParameter(const std::string param_name, const rclcpp::ParameterValue paramvalue)
+{
+  RTT::Logger::In in(getName());
+
+  RTT::log(RTT::Debug) <<
+    "Setting the parameter \"" << param_name << "\"" <<
+    RTT::endlog();
+
+  rclcpp::Node::SharedPtr rosnode;
+  if (get_ros2_node(rosnode)) {
+    if (!rosnode->has_parameter(param_name)) {
+      // Create new parameter
+      RTT::log(RTT::Warning) <<
+        "The parameter did not exist, creating one" <<
+        RTT::endlog();
+      rosnode->declare_parameter(param_name);
+    }
+
+    // Update the parameter
+    try {
+      const auto result = rosnode->set_parameter(rclcpp::Parameter(param_name, paramvalue));
+      if (!result.successful) {
+        RTT::log(RTT::Error) <<
+          "The parameter " << param_name <<
+          " couldn't be set with reason: " << result.reason << RTT::endlog();
+        return false;
+      }
+    } catch (std::exception & e) {
+      RTT::log(RTT::Error) <<
+        "The parameter failed to be set, reason: " <<
+        e.what() << RTT::endlog();
+    }
+  } else {
+    RTT::log(RTT::Error) <<
+      "The ROS2 node doesn't exist, no parameter can be set. Import rtt_ros2_node first and use "
+      "the \"ros2-node\" service." <<
+      RTT::endlog();
+    return false;
+  }
+  return true;
+}
+
+bool Params::loadProperty(const std::string param_name, const std::string property_name)
+{
+  RTT::Logger::In in(getName());
+
+  rclcpp::ParameterValue paramvalue = getParameter(param_name);
+  if (rclcpp::PARAMETER_NOT_SET == paramvalue.get_type()) {
+    RTT::log(RTT::Error) <<
+      "The parameter \"" << param_name << "\" couldn't be retrieved" <<
+      RTT::endlog();
+    return false;
+  }
+
+  if (nullptr != getOwner()->properties()->getProperty(property_name)) {
+    RTT::log(RTT::Debug) <<
+      getOwner()->provides()->getName() <<
+      " has a property " << property_name << RTT::endlog();
+  } else {
+    RTT::log(RTT::Info) <<
+      getOwner()->provides()->getName() <<
+      " has NO property " << property_name << RTT::endlog();
+    for (const auto property : getOwner()->properties()->getProperties() ) {
+      RTT::log(RTT::Debug) <<
+        "Available property: " << property->getName() <<
+        RTT::endlog();
+    }
+  }
+
+  auto prop = getOwner()->properties()->getProperty(property_name);
+  if (nullptr == prop) {
+    // Try to find it among orphan parameters (previously loaded, without a component interface)
+    if (orphan_params_.find(property_name) == orphan_params_.end()) {
+      // When totally new, then create an orphan property, i.e. owned by the ros2-params service
+      orphan_params_[property_name] = paramvalue;
+      this->addProperty(property_name, orphan_params_[property_name])
+      .doc("Property loaded from ROS2 param: " + param_name);
+      prop = getProperty(property_name);
+    } else {
+      orphan_params_[property_name] = paramvalue;
+      prop = getProperty(property_name);
+    }
```

Comment:
Can you explain the use case and what is meant by "orphan(ed) parameters"?

I don't think that we should add properties to the `params` service itself, always to the owning `TaskContext` (and do not expose this method if the service does not have an owner, i.e. for the global service). At least the behavior of `loadProperty()` should be consistent and only operate on existing properties. There could be other operations that allow to create new properties, like `loadOrAddProperty()`.

In any case new properties being added to either service should be of the actual underlying type (e.g. int, string, bool, ...) and not of type `ParameterValue`.

Reply:
The problem of adding this orphan(ed) properties (not parameters, I changed that) is that the `TaskContext` does not foresee memory for these properties. So I created the map `std::map<std::string, rclcpp::ParameterValue> orphan_properties_;` in the service itself.
I am not sure how to otherwise make the TC owner of these new properties.
Also, if not stored as `ParameterValue`, I need to create one map as per possible type.

I agree on giving this functionality only to another operation like `loadOrAddProperty()`, since the use-case might be rare. Such example could be a component that logs what is loaded into the ROS interface.